### ISSUE TITLE
Timeouts when handling messages containing many lookups

### ISSUE THREAD

**Issue by OliverHofkens** [OliverHofkens/scryfall-telegram-rs-serverless#4]
Judging from the live Lambda monitoring, people are using this bot to share entire decklists in chat. This is currently causing timeouts, as we can't complete 40+ lookups through the Scryfall API in under 10 seconds. 
I'm also guessing this is gonna get us rate-limited or banned from the Scryfall API pretty quickly.

2 questions need to be handled to tackle this issue:
- Should we limit the amount of lookups in a single message, or can we do multiple lookups in a single API call perhaps?
- How do we visualize a 60+ card gallery in a Telegram chat?

**Comment by suroh1994** [comment]
You could process the request asynchronously by sending a placeholder message like "Retrieving images: 5/60" and update it while you're processing the request. Once you have all pictures gathered, you can sent them at once and delete the original message. Though that would mean the message might be delayed by a more significant time. 
If you create a central queue for all request to scryfall, you could do some rate throttling to keep the bot from being banned.

Just some ideas, though I personally don't see why you would share a whole decklist as images through telegram. But then again, if there is a significant user base doing so, it's up to you to decide whether you're going to support it. I'd say a max of maybe 10 Cards should be more than enough for a single message.

**Comment by henges** [comment]
I am on the side of limiting the amount of lookups in a single message (to 20 or thereabouts). There are many established online tools for sharing decklists that are better suited to that purpose anyway (they can categorise and sort them etc). In my experience, the utility of a bot like this one is being able to immediately have a few cards' text available in a group chat for everyone to see so that you can discuss them without having to leave the application. 
Telegram also doesn't cope well with sending many images at once. The last time I tried to send a large amount of photos (~100) at once it automatically split them into groups of 10, although that might have changed recently.
Alternatively, if you did want to support sending many images, I see the [API documentation](https://scryfall.com/docs/api) recommends a rate of about 10 requests per second, so perhaps the bot could space the requests appropriately if the number of requests is greater than 10.

**Comment by OliverHofkens** [comment]
Those are good points and valid suggestions, thanks!

**Comment by OliverHofkens** [comment]
Telegram's MediaGallery is limited to 10 items, so I limited the bot to fetch 10 items as well. 
I also increased the timeout from 10s to 30s, as it's better to wait a bit longer once then to let the bot timeout and retry over and over again.
I didn't look into spacing the requests yet as I noticed that the response time of Scryfall is always over the recommended 100ms, and since the bot is synchronous, the requests should actually respect the rate limit already.

Thanks again for your input!
